feat: add C `ndarray` implementation for `blas/base/zscal`

aman-095 opened the pull request:
Progresses #2039.

## Description

> What is the purpose of this pull request?

This pull request adds C `ndarray` implementation for `zscal`. 

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aman-095 committed:
feat: add C ndarray implementation for zscal

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/blas/base/zscal/lib/ndarray.js:
```diff
@@ -48,35 +45,17 @@ var imag = require( '@stdlib/complex/float64/imag' );
 		return zx;
 	}
-	// Reinterpret the input array as a real-valued array of interleaved real and imaginary components:
-	view = reinterpret( zx, 0 );
```
@aman-095 Please revert this refactoring. I just recently updated this to avoid complex number instantiation. Don't simply copy-paste from `cscal`.

kgryte commented on lib/node_modules/@stdlib/blas/base/zscal/lib/zscal.js:
```diff
@@ -32,7 +32,7 @@ var ndarray = require( './ndarray.js' );
 * @param {Complex128} za - constant
 * @param {Complex128Array} zx - input array
-* @param {integer} strideZX - `zx` stride length
+* @param {integer} strideX - `zx` stride length
```
FTR, I recently changed these: https://github.com/stdlib-js/stdlib/commit/c97212e122c2d20800469913282d9df791fcebf2#diff-1135e703968b45bc6a897ad9dc728fb0d90a6b633a520e9c455d7bc6c13a78caR36

Namely, because we are not consistent. Personally, instead of `za` and `zx`, I'd prefer if we just used `alpha` and `x` for both real and complex. The use of `za`, `ca`, etc, stems from Fortran in order to signal the dtype. I think we just create more work for ourselves when going from real to complex by following suit and renaming variables from `x` to `cx` or `zx`. The function name and the C type signatures are enough to convey what is expected. I don't think we need to encode the type info directly into the parameter names themselves.

aman-095 replied:
I agree, so should I change the parameter names from `zx` -> `x` and `za` -> `alpha`.

kgryte replied:
Yes, that works for me. Thanks!

kgryte commented on lib/node_modules/@stdlib/blas/base/zscal/test/test.ndarray.js:
```diff
@@ -138,26 +154,60 @@ tape( 'the function supports a `zx` stride', function test( t ) {
+	t.end();
+});
+
+tape( 'the function supports a negative `zx` stride', function test( t ) {
```
@aman-095 Why is this added given L250 which tests the same thing?

aman-095 replied:
Updated.

kgryte reviewed (changes requested):
Left an initial round of comments.

aman-095 committed:
refactor: revert implementation for ndarray

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aman-095 committed:
test: remove duplicate tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

aman-095 committed:
Merge branch 'stdlib-js:develop' into zscal-ndarray

aman-095 committed:
chore: update description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aman-095 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into zscal-ndarray

aman-095 committed:
chore: resolve lint error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte committed:
Merge branch 'develop' into zscal-ndarray

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zscal/README.md:
```diff
@@ -127,7 +127,7 @@ zscal.ndarray( 2, za, zx, 2, 1 );
 ## Notes
 
--   If `N <= 0` or `strideX <= 0` , both functions return `zx` unchanged.
+-   If `N <= 0` or `strideX <= 0` , both functions return `x` unchanged.
```
```suggestion
-   If `N <= 0` , both functions return `x` unchanged.
```

kgryte committed:
docs: fix note

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zscal/docs/repl.txt:
```diff
@@ -1,15 +1,15 @@
 
-    The `N` and stride parameters determine how values from `zx` are scaled by
-    `za`.
+    The `N` and stride parameters determine how values from `x` are scaled by
```
```suggestion
    The `N` and stride parameters determine which values from `x` are scaled by
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/zscal/docs/repl.txt:
```diff
@@ -1,15 +1,15 @@
     array views.
 
-    If `N` or `strideX` is less than or equal to `0`, the function returns `zx`
+    If `N` or `strideX` is less than or equal to `0`, the function returns `x`
     unchanged.
```
```suggestion
    If `N` is less than or equal to `0`, the function returns `x` unchanged.
```

kgryte committed:
docs: update note

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM

kgryte commented on lib/node_modules/@stdlib/blas/base/zscal/docs/repl.txt:
```diff
@@ -1,62 +1,61 @@
-    > var za = new {{alias:@stdlib/complex/float64/ctor}}( 2.0, 2.0 );
-    > {{alias}}( 2, za, zx1, 1 )
+    > var x0 = new {{alias:@stdlib/array/complex128}}( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
+    > var x1 = new {{alias:@stdlib/array/complex128}}( zx0.buffer, zx0.BYTES_PER_ELEMENT*1 );
```
```suggestion
    > var x1 = new {{alias:@stdlib/array/complex128}}( x0.buffer, x0.BYTES_PER_ELEMENT*1 );
```

kgryte committed:
docs: fix example

Signed-off-by: Athan <kgryte@gmail.com>
